.github/actions/yarn-nm-install/action.yml

```diff
@@ -0,0 +1,65 @@
+########################################################################################
+# "yarn install" composite action for yarn 3/4+ and "nodeLinker: node-modules"         #
+#--------------------------------------------------------------------------------------#
+# References:                                                                          #
+#   - bench: https://gist.github.com/belgattitude/0ecd26155b47e7be1be6163ecfbb0f0b     #
+########################################################################################
+
+name: 'Monorepo install (yarn)'
+description: 'Run yarn install with node_modules linker and cache enabled'
+inputs:
+  enable-corepack:
+    description: 'Enable corepack'
+    required: false
+    default: 'false'
+  cache-install-state:
+    description: 'Cache yarn install state (invalidated on lock changes)'
+    required: false
+    default: 'true'
+  cache-node-modules:
+    description: 'Cache node_modules (invalidated on lock changes)'
+    required: false
+    default: 'false'
+
+runs:
+  using: 'composite'
+
+  steps:
+    - name: ⚙️ Enable Corepack
+      if: ${{ inputs.enable-corepack }} == 'true'
+      shell: bash
+      run: corepack enable
+
+    - name: ⚙️ Expose yarn config as "$GITHUB_OUTPUT"
+      id: yarn-config
+      shell: bash
+      run: |
+        echo "CACHE_FOLDER=$(yarn config get cacheFolder)" >> $GITHUB_OUTPUT
+
+    - name: ♻️ Restore yarn cache
```

Comment:
That will store/restore the `.yarn/cache/*.zip` files. They are required even if you have node_modules. The good thing they are platform independentent (except if using supportedArchitectures in yarn 4). Another great feature, yarn manages pruning (does not need to be invalidated on lock changes). It wont grow, and yarn can start without refetching everything